src/platform/graphics/blend-state.js

```diff
@@ -0,0 +1,222 @@
+     * Defaults to true.
+     * @param {boolean} greenWrite - True to enable writing of the green channel and false
+     * otherwise. Defaults to true.
+     * @param {boolean} blueWrite - True to enable writing of the blue channel and false otherwise.
+     * Defaults to true.
+     * @param {boolean} alphaWrite - True to enable writing of the alpha channel and false
+     * otherwise. Defaults to true.
+     */
+    constructor(blend = false, colorOp = BLENDEQUATION_ADD, colorSrcFactor = BLENDMODE_ONE, colorDstFactor = BLENDMODE_ZERO,
+        alphaOp, alphaSrcFactor, alphaDstFactor,
+        redWrite = true, greenWrite = true, blueWrite = true, alphaWrite = true) {
+        this.setColorBlend(colorOp, colorSrcFactor, colorDstFactor);
+        this.setAlphaBlend(alphaOp ?? colorOp, alphaSrcFactor ?? colorSrcFactor, alphaDstFactor ?? colorDstFactor);
+        this.setColorWrite(redWrite, greenWrite, blueWrite, alphaWrite);
+        this.blend = blend;
+    }
+
+    /**
+     * Enables or disables blending.
+     *
+     * @type {boolean}
+     */
+    set blend(value) {
+        this.target0 = BitPacking.set(this.target0, value ? 1 : 0, blendShift);
+    }
+
+    get blend() {
+        return BitPacking.all(this.target0, blendShift);
+    }
+
+    setColorBlend(op, srcFactor, dstFactor) {
+        this.target0 = BitPacking.set(this.target0, op, colorOpShift, opMask);
+        this.target0 = BitPacking.set(this.target0, srcFactor, colorSrcFactorShift, factorMask);
+        this.target0 = BitPacking.set(this.target0, dstFactor, colorDstFactorShift, factorMask);
+    }
+
+    setAlphaBlend(op, srcFactor, dstFactor) {
+        this.target0 = BitPacking.set(this.target0, op, alphaOpShift, opMask);
+        this.target0 = BitPacking.set(this.target0, srcFactor, alphaSrcFactorShift, factorMask);
+        this.target0 = BitPacking.set(this.target0, dstFactor, alphaDstFactorShift, factorMask);
+    }
+
+    setColorWrite(redWrite, greenWrite, blueWrite, alphaWrite) {
+        this.redWrite = redWrite;
+        this.greenWrite = greenWrite;
+        this.blueWrite = blueWrite;
+        this.alphaWrite = alphaWrite;
+    }
+
+    get colorOp() {
+        return BitPacking.get(this.target0, colorOpShift, opMask);
+    }
+
+    get colorSrcFactor() {
+        return BitPacking.get(this.target0, colorSrcFactorShift, factorMask);
+    }
+
+    get colorDstFactor() {
+        return BitPacking.get(this.target0, colorDstFactorShift, factorMask);
+    }
+
+    get alphaOp() {
+        return BitPacking.get(this.target0, alphaOpShift, opMask);
+    }
+
+    get alphaSrcFactor() {
+        return BitPacking.get(this.target0, alphaSrcFactorShift, factorMask);
+    }
+
+    get alphaDstFactor() {
+        return BitPacking.get(this.target0, alphaDstFactorShift, factorMask);
+    }
+
+    set redWrite(value) {
+        this.target0 = BitPacking.set(this.target0, value ? 1 : 0, redWriteShift);
```

Comment:
Either what I said above, or these should also set the redWrite, greenWrite, etc.

Reply:
I'm not sure I follow. All data is stored in target0 at the moment, including redWrite and all that. redWrite is just a setter that does it.